File: web/status-history/src/test/java/net/twisterrob/travel/statushistory/viewmodel/GwenChange.java
package net.twisterrob.travel.statushistory.viewmodel;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.junit.Assert.*;

import com.shazam.gwen.collaborators.*;

import net.twisterrob.blt.model.Line;
import net.twisterrob.travel.statushistory.viewmodel.ResultChange.ErrorChange;
import net.twisterrob.travel.statushistory.viewmodel.ResultChange.StatusChange;

class GwenChange implements Actor, Asserter {
	private ResultChange change;

	public void between(GwenStatus status1, GwenStatus status2) {
		change = new ResultChange(status1.createResult(), status2.createResult());
	}

	public GwenChange has(Line line, StatusChange status) {
		assertThat(change.getStatuses(), hasEntry(line, status));
		return this;
	}

	public GwenChange hasNoDescriptionChangeFor(Line... lines) {
		for (Line line : lines) {
			assertThat(change.getDescriptions(), not(hasKey(line)));
		}
		return this;
	}

	public GwenChange hasDescriptionChangeFor(Line... lines) {
		for (Line line : lines) {
			assertThat(change.getDescriptions(), hasEntry(equalTo(line), not(emptyOrNullString())));
		}
		return this;
	}

	public GwenChange hasNoErrorChange() {
		assertEquals(ErrorChange.NoErrors, change.getError());
		return this;
	}
}

File: web/status-history/src/test/java/net/twisterrob/travel/statushistory/viewmodel/GwenChange.kt
package net.twisterrob.travel.statushistory.viewmodel

import com.shazam.gwen.collaborators.Actor
import com.shazam.gwen.collaborators.Asserter
import net.twisterrob.blt.io.feeds.trackernet.model.DelayType
import net.twisterrob.blt.model.Line
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.either
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.hasEntry
import org.hamcrest.Matchers.hasKey
import org.hamcrest.Matchers.instanceOf
import org.hamcrest.Matchers.not

internal class GwenChange : Actor, Asserter {

	private lateinit var difference: Changes

	private val statusChanges: Map<Line, StatusChange>
		get() = (difference as Changes.Status).changes

	fun between(status1: GwenStatus, status2: GwenStatus) {
		difference = ResultChangesCalculator().diff(status1.createResult(), status2.createResult())
	}

	fun has(line: Line, desc: DescriptionChange): GwenChange = apply {
		assertThat(statusChanges, hasEntry(line, StatusChange.Same(DelayType.Unknown, desc)))
	}

	fun has(line: Line, status: StatusChange): GwenChange = apply {
		assertThat(statusChanges, hasEntry(line, status))
	}

	fun hasNoStatusChangeFor(vararg lines: Line): GwenChange = apply {
		for (line in lines) {
			assertThat(statusChanges, not(hasKey(line)))
		}
	}

	fun hasNoDescriptionChangeFor(vararg lines: Line): GwenChange = apply {
		for (line in lines) {
			assertThat(statusChanges, either(not(hasKey(line))).or(not(instanceOf(HasDescriptionChange::class.java))))
		}
	}

	fun hasDescriptionChangeFor(vararg lines: Line): GwenChange = apply {
		for (line in lines) {
			assertThat(statusChanges, hasEntry(equalTo(line), instanceOf(HasDescriptionChange::class.java)))
		}
	}

	fun hasNoErrorChange(): GwenChange = apply {
		assertThat(difference, instanceOf(Changes.Status::class.java))
	}
}
